resolwe/flow/models.py

```diff
@@ -384,26 +391,40 @@ class Meta(BaseModel.Meta):
     #: actual descriptor
     descriptor = JSONField(default={})
 
+    # track if user set the data name explicitly
+    named_by_user = models.BooleanField(default=False)
+
+    def __init__(self, *args, **kwargs):
+        super(Data, self).__init__(*args, **kwargs)
+        self._original_name = self.name
+
     def save(self, *args, **kwargs):
         # Generate the descriptor if one is not already set.
         if not self.descriptor:
             render_descriptor(self)
 
-        if not self.name:
-            render_name(self)
+        if self.name != self._original_name:
+            self.named_by_user = True
 
-        created = False
-        if not self.pk:  # create
-            created = True
+        create = self.pk is None
+        if create:
+            if not self.name:
+                render_name(self)
+            else:
+                self.named_by_user = True
 
-            # default values for INPUT
+            # Default values for INPUT
             for field_schema, fields, path in iterate_schema(self.input, self.process.input_schema, ''):
                 if 'default' in field_schema and field_schema['name'] not in fields:
                     dict_dot(self.input, path, field_schema['default'])
 
+        elif kwargs.get('render_name', False):
+            render_name(self)
+            del kwargs['render_name']
```

Comment:
Define `save` function as `save(self, render_name=False, *args, **kwargs):` and it will be a lot nicer.

Django [`save` function](https://docs.djangoproject.com/en/1.9/_modules/django/db/models/base/#Model.save) already has few such arguments